src/domain/payments/onchain-payment-flow-builder.ts

```diff
@@ -0,0 +1,571 @@
+import { getFeesConfig } from "@config"
+import {
+  AmountCalculator,
+  paymentAmountFromNumber,
+  ValidationError,
+  WalletCurrency,
+  ZERO_BANK_FEE,
+} from "@domain/shared"
+import { LessThanDustThresholdError, SelfPaymentError } from "@domain/errors"
+import { OnChainFees, PaymentInitiationMethod, SettlementMethod } from "@domain/wallets"
+import { checkedToBtcPaymentAmount, checkedToUsdPaymentAmount } from "@domain/payments"
+import { ImbalanceCalculator } from "@domain/ledger/imbalance-calculator"
+
+import { InvalidOnChainPaymentFlowBuilderStateError } from "./errors"
+import { PriceRatio } from "./price-ratio"
+import { OnChainPaymentFlow } from "./payment-flow"
+
+const calc = AmountCalculator()
+const feeConfig = getFeesConfig()
+const onChainFees = OnChainFees({
+  feeRatioAsBasisPoints: feeConfig.withdrawRatioAsBasisPoints,
+  thresholdImbalance: {
+    amount: BigInt(feeConfig.withdrawThreshold),
+    currency: WalletCurrency.Btc,
+  },
+})
+
+export const OnChainPaymentFlowBuilder = <S extends WalletCurrency>(
+  config: OnChainPaymentFlowBuilderConfig,
+): OnChainPaymentFlowBuilder<S> => {
+  const withAddress = (address: OnChainAddress): OPFBWithAddress<S> | OPFBWithError => {
+    // TODO: validate onchain address?
```

Comment:
The type `OnChainAddress` implies its already been validates (or it would be `string` type right?)